fix local_group::erase_subscriber

nq-ebaratte opened the pull request:
Fixes #574.

This PR makes use of a functionnality introduced in C++14: the ability to use a type that "compares equivalent" to the keys of an` std::set<>` in its `find()` method.

I added the `operator<` to compare an `intrusive_ptr<T>` to a raw `T*`

As an added bonus, the deletion is now O(log(N)), instead of O(N)

Neverlord commented:
It's too bad, but I have to turn this PR down. CAF uses C++11, because we still support GCC 4.8. I look forward to drop support for this relatively old compiler eventually, but now isn't the time unfortunately.

Neverlord commented:
Thinking a bit more about it: we shouldn't discard your work entirely. Can you surround the critical code piece with an `#if __cplusplus > 201103L` guard and provide a C++11 fallback?

nq-ebaratte commented:
I updated the PR.

Please note that I dropped the actor_addr::compare() test, and use only address comparison. Is there a case where two different actor_control_block would describe the same actor ?

Neverlord commented:
Thanks a lot! 👍🏻

Neverlord commented:
Sorry, I forgot to answer your question.

> Is there a case where two different actor_control_block would describe the same actor?

For local actors: no.

I think there is only one way this could happen in CAF: storing an `actor_addr` to a remote actor. If the remote actor terminates, all actors receive down/exit messages and after that point CAF can drop the handle to it from the proxy registry. If we receive a message with a handle to the terminated actor then we re-create the handle. However, sending messages to the newly created handle will have no effect (actor already dead). Having the old weak pointer laying around therefore is not an issue, as you need to cast it to a strong pointer before doing anything with it (which will fail).

Relevant to this PR: the group only stores handles to local actors. So there's no way two control blocks identify the same actor.
